# Soccer Data
This dataset contains data of every game from the 2018-2019 season in the English Premier League. 


In [2]:
#!/usr/bin/env python3


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.read_csv("soccer_data.csv")
soccer_data = pd.read_csv("soccer_data.csv")

## Data Dictionary
| Column    | Explanation                                   |
| --------- | --------------------------------------------- |
| Div       | Division the game was played in               |                            
| Date      | The date the game was played                  |
| HomeTeam  | The home team                                 |
| AwayTeam  | The away team                                 |
| FTHG      | Full time home goals                          |
| FTAG      | Full time away goals                          |
| FTR       | Full time result                              |
| HTHG      | Half time home goals                          |
| HTAG      | Half time away goals                          |
| HTR       | Half time result                              |
| Referee   | The referee of the game                       |
| HS        | Number of shots taken by home team            |
| AS        | Number of shots taken by away team            |
| HST       | Number of shots taken by home team on target  |
| AST       | Number of shots taken by away team on target  |
| HF        | Number of fouls made by home team             |
| AF        | Number of fouls made by away team             |
| HC        | Number of corners taken by home team          |
| AC        | Number of corners taken by away team          |
| HY        | Number of yellow cards received by home team  |
| AY        | Number of yellow cards received by away team  |
| HR        | Number of red cards received by home team     |
| AR        | Number of red cards received by away team     |

In [3]:
soccer_data.describe()

FTHG        FTAG        HTHG        HTAG          HS          AS  \
count  380.000000  380.000000  380.000000  380.000000  380.000000  380.000000   
mean     1.568421    1.252632    0.678947    0.573684   14.134211   11.144737   
std      1.312836    1.180031    0.860802    0.766958    5.855371    4.654002   
min      0.000000    0.000000    0.000000    0.000000    0.000000    2.000000   
25%      1.000000    0.000000    0.000000    0.000000   10.000000    8.000000   
50%      1.000000    1.000000    0.000000    0.000000   14.000000   11.000000   
75%      2.000000    2.000000    1.000000    1.000000   18.000000   14.000000   
max      6.000000    6.000000    4.000000    3.000000   36.000000   25.000000   

              HST         AST          HF          AF          HC          AC  \
count  380.000000  380.000000  380.000000  380.000000  380.000000  380.000000   
mean     4.778947    3.928947   10.152632   10.305263    5.705263    4.552632   
std      2.677686    2.283982    3.293532    3.503707    2.971718    2.730627   
min      0.000000    0.000000    0.000000    3.000000    0.000000    0.000000   
25%      3.000000    2.000000    8.000000    8.000000    4.000000    2.750000   
50%      5.000000    4.000000   10.000000   10.000000    5.000000    4.000000   
75%      6.000000    5.250000   12.000000   13.000000    8.000000    6.000000   
max     14.000000   12.000000   23.000000   21.000000   16.000000   14.000000   

               HY          AY          HR          AR  
count  380.000000  380.000000  380.000000  380.000000  
mean     1.526316    1.684211    0.047368    0.076316  
std      1.222844    1.209140    0.212706    0.275599  
min      0.000000    0.000000    0.000000    0.000000  
25%      1.000000    1.000000    0.000000    0.000000  
50%      1.000000    2.000000    0.000000    0.000000  
75%      2.000000    2.000000    0.000000    0.000000  
max      6.000000    5.000000    1.000000    2.000000

In [4]:
soccer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380 entries, 0 to 379
Data columns (total 23 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Div       380 non-null    object
 1   Date      380 non-null    object
 2   HomeTeam  380 non-null    object
 3   AwayTeam  380 non-null    object
 4   FTHG      380 non-null    int64 
 5   FTAG      380 non-null    int64 
 6   FTR       380 non-null    object
 7   HTHG      380 non-null    int64 
 8   HTAG      380 non-null    int64 
 9   HTR       380 non-null    object
 10  Referee   380 non-null    object
 11  HS        380 non-null    int64 
 12  AS        380 non-null    int64 
 13  HST       380 non-null    int64 
 14  AST       380 non-null    int64 
 15  HF        380 non-null    int64 
 16  AF        380 non-null    int64 
 17  HC        380 non-null    int64 
 18  AC        380 non-null    int64 
 19  HY        380 non-null    int64 
 20  AY        380 non-null    int64 
 21  HR        380 no

In [5]:
soccer_data.head()

Div     Date      HomeTeam        AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  \
0  E0  10/8/18    Man United       Leicester     2     1   H     1     0   H   
1  E0  11/8/18   Bournemouth         Cardiff     2     0   H     1     0   H   
2  E0  11/8/18        Fulham  Crystal Palace     0     2   A     0     1   A   
3  E0  11/8/18  Huddersfield         Chelsea     0     3   A     0     2   A   
4  E0  11/8/18     Newcastle       Tottenham     1     2   A     1     2   A   

   ... HST  AST  HF  AF  HC  AC  HY  AY  HR  AR  
0  ...   6    4  11   8   2   5   2   1   0   0  
1  ...   4    1  11   9   7   4   1   1   0   0  
2  ...   6    9   9  11   5   5   1   2   0   0  
3  ...   1    4   9   8   2   5   2   1   0   0  
4  ...   2    5  11  12   3   5   2   2   0   0  

[5 rows x 23 columns]

In [6]:
# check nulls
soccer_data.notnull().sum()

Div         380
Date        380
HomeTeam    380
AwayTeam    380
FTHG        380
FTAG        380
FTR         380
HTHG        380
HTAG        380
HTR         380
Referee     380
HS          380
AS          380
HST         380
AST         380
HF          380
AF          380
HC          380
AC          380
HY          380
AY          380
HR          380
AR          380
dtype: int64

In [7]:
corners_home = soccer_data['HC']

## What team commits the most fouls?

First, we need to know how many matches every team played. 

Let's assume that it will be the same number for every team - 38 matches - 19 at home, and 19 away.

In [8]:
count_games = soccer_data['HomeTeam'].str.count('Liverpool').sum()

sum_home_fouls = soccer_data.groupby(['HomeTeam'])['HF'].sum()

sum_away_fouls = soccer_data.groupby(['AwayTeam'])['AF'].sum()

fouls = pd.concat([sum_home_fouls, sum_away_fouls], axis=1)
fouls['Sum_fouls'] = fouls.apply(lambda row: row.HF + row.AF, axis=1)
fouls['Average_per_match'] = fouls.apply(lambda row: row.Sum_fouls / 38, axis=1)
fouls

HF   AF  Sum_fouls  Average_per_match
Arsenal         218  194        412          10.842105
Bournemouth     187  152        339           8.921053
Brighton        221  242        463          12.184211
Burnley         184  175        359           9.447368
Cardiff         195  189        384          10.105263
Chelsea         154  185        339           8.921053
Crystal Palace  180  206        386          10.157895
Everton         218  214        432          11.368421
Fulham          199  224        423          11.131579
Huddersfield    214  206        420          11.052632
Leicester       184  167        351           9.236842
Liverpool       149  166        315           8.289474
Man City        162  166        328           8.631579
Man United      213  220        433          11.394737
Newcastle       203  219        422          11.105263
Southampton     219  201        420          11.052632
Tottenham       174  201        375           9.868421
Watford         221  212        433          11.394737
West Ham        165  166        331           8.710526
Wolves          198  211        409          10.763158

In [9]:
fouls['Sum_fouls'].describe()

count     20.000000
mean     388.700000
std       43.785482
min      315.000000
25%      348.000000
50%      397.500000
75%      422.250000
max      463.000000
Name: Sum_fouls, dtype: float64

In [10]:
fouls['Average_per_match'].describe()

count    20.000000
mean     10.228947
std       1.152250
min       8.289474
25%       9.157895
50%      10.460526
75%      11.111842
max      12.184211
Name: Average_per_match, dtype: float64

With an average of 12 fouls per match, Brighton players committed the most fouls (season - 463).

More than half of the teams committed an average of ten fouls.

Liverpool players committed the fewest fouls, with an average of 8 fouls per game, which was less than the league average.

In [11]:
fouls_sorted = fouls.sort_values('Average_per_match', ascending=False)
fouls_sorted[['Average_per_match']].iloc[:5]

Average_per_match
Brighton            12.184211
Watford             11.394737
Man United          11.394737
Everton             11.368421
Fulham              11.131579

In [12]:
fouls_sorted[['Average_per_match']].iloc[-1]

Average_per_match    8.289474
Name: Liverpool, dtype: float64

##  Which referee handled the majority of the matches?

The majority of the matches were officiated by Antony Taylor, who issued 102 yellow cards to players, but only one red card. 

In [13]:
ref_list = list(soccer_data['Referee'].unique())
print('List of referees in season: ')
for ref in ref_list: print(f'- {ref}')

List of referees in season: 
- A Marriner
- K Friend
- M Dean
- C Kavanagh
- M Atkinson
- J Moss
- C Pawson
- M Oliver
- A Taylor
- G Scott
- L Mason
- S Attwell
- P Tierney
- L Probert
- D Coote
- R East
- S Hooper
- A Madley


In [14]:
soccer_data['Referee'].value_counts()

A Taylor      32
M Oliver      30
M Dean        29
M Atkinson    29
A Marriner    27
J Moss        27
K Friend      27
C Pawson      26
C Kavanagh    24
P Tierney     24
S Attwell     20
L Mason       19
L Probert     18
G Scott       17
D Coote       11
R East        10
S Hooper       8
A Madley       2
Name: Referee, dtype: int64

In [15]:
df = soccer_data[['Referee', 'HY', 'AY']]
df_index = df.set_index('Referee')
df_x = df_index.groupby('Referee')[['HY', 'AY']].sum()
df_x['card_sum'] = df_x.apply(lambda row: row.HY + row.AY, axis=1)
df_x.sort_values('card_sum',ascending=False)

HY  AY  card_sum
Referee                     
M Dean      62  65       127
A Taylor    57  45       102
J Moss      44  49        93
C Pawson    46  46        92
K Friend    48  44        92
M Atkinson  38  53        91
M Oliver    35  48        83
P Tierney   37  45        82
C Kavanagh  39  35        74
A Marriner  31  42        73
L Mason     30  33        63
S Attwell   31  32        63
L Probert   20  26        46
R East      23  21        44
G Scott     14  24        38
D Coote     16  19        35
S Hooper     9  11        20
A Madley     0   2         2

In [16]:
x = soccer_data.groupby('Referee')[['HR', 'AR']].sum()
x['cards_sum_red'] = x.apply(lambda row: row.HR + row.AR, axis=1)
x.sort_values('cards_sum_red',ascending=False)

HR  AR  cards_sum_red
Referee                          
M Dean       3   7             10
C Pawson     1   6              7
M Oliver     3   4              7
J Moss       3   2              5
L Probert    1   3              4
K Friend     2   1              3
A Marriner   2   1              3
C Kavanagh   1   1              2
L Mason      1   0              1
A Taylor     0   1              1
S Attwell    0   1              1
G Scott      0   1              1
M Atkinson   0   1              1
P Tierney    1   0              1
R East       0   0              0
A Madley     0   0              0
D Coote      0   0              0
S Hooper     0   0              0